Mirror/Runtime/NetworkBehaviour.cs

```diff
@@ -2,12 +2,20 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using UnityEngine;
+using System.Linq;
 
 namespace Mirror
 {
+    // provide an interface so that we can mock it in unit tests
+    public interface INetworkBehaviour
+    {
+        bool isServer { get; }
+        bool isClient { get; }
+    }
+
```

Comment:
can you explain the interface part please? which tests do we need it for and why do we need it?

Reply:
Yes,  it works this way:

I want to unit test synclists.   But synclists have a dependency on NetworkBehavior.  This dependency was there in HLAPI and it is used to determine when the synclist should call the callback.    

The problem is that unit tests cannot instantiate NetworkBehaviors  (or any other MonoBehavior).  So how do we test synclists without instantiating NetworkBehaviors?

The solution is the well known and best practice Interface Segregation Principle https://en.wikipedia.org/wiki/Interface_segregation_principle.   

Instead of receiving a NetworkBehavior,  synclists now receive any object that has an isServer and isClient method,  because that is really all they need.  This is done by depending on the interface instead of directly on NetworkBehavior.  NetworkBehavior happens to be an implementation of INetworkBehavior.

The tests can now create a fake NetworkBehavior that implements that interface instead of a real NetworkBehavior and pass it to the synclists prior to testing them.   That is well known method for unit testing called mocking.   NSubstitute is the most popular mocking framework for C#,   you can create a mock for any interface or delegate and use it to fulfill dependencies.  

The Interface Segregation Principle is the I in SOLID principles. It is well known,  best practice in OOP and unity in general does not do enough of it,  especially since they are piss poor at testing.

Reply:
Okay, I can see the value in it. 
Well, honestly I think it's too complicated. I like tests a lot but I don't think we should make the code more complicated for the tests. There is so much magic in hlapi/weaver already.

Any chance we could see a version without the interfaces (even if it means no tests) and possibly a version with the alternative way that you explained in the end? Unless that's more complicated too.

Sorry, I have a small brain, need to keep it simple :)